src/user-statuses/__tests__/userStatusReducer-test.js

```diff
@@ -0,0 +1,117 @@
+/* @flow strict-local */
+import Immutable from 'immutable';
+
+import { makeUserId } from '../../api/idTypes';
+import type { UserStatusesState } from '../../types';
+import * as eg from '../../__tests__/lib/exampleData';
+import { EVENT_USER_STATUS_UPDATE } from '../../actionConstants';
+import userStatusesReducer from '../userStatusesReducer';
+
+describe('userStatusesReducer', () => {
+  const testUserStatusState: UserStatusesState = Immutable.Map([
+    [makeUserId(1), { away: true, status_text: null, status_emoji: null }],
+    [makeUserId(2), { away: false, status_text: 'Hello, world', status_emoji: null }],
+  ]);
+
+  describe('ACCOUNT_SWITCH', () => {
+    test('resets state to initial state', () => {
+      expect(userStatusesReducer(testUserStatusState, eg.action.account_switch)).toEqual(
+        Immutable.Map(),
+      );
+    });
+  });
+
+  describe('REGISTER_COMPLETE', () => {
+    test('when `user_status` data is provided init state with it', () => {
+      expect(
+        userStatusesReducer(
+          Immutable.Map(),
+          eg.mkActionRegisterComplete({
+            user_status: { '1': { away: true }, '2': { status_text: 'Hello, world' } },
+          }),
+        ),
+      ).toEqual(testUserStatusState);
+    });
+
+    test('handles older back-ends that do not have `user_status` data by resetting the state', () => {
+      expect(
+        userStatusesReducer(
+          testUserStatusState,
+          eg.mkActionRegisterComplete({ user_status: undefined }),
+        ),
+      ).toEqual(Immutable.Map());
+    });
+  });
+
+  describe('EVENT_USER_STATUS_UPDATE', () => {
+    test('when the user does not already have entry add a key by their `user_id`', () => {
+      expect(
+        userStatusesReducer(testUserStatusState, {
+          id: 0,
+          type: EVENT_USER_STATUS_UPDATE,
+          user_id: makeUserId(3),
+          away: true,
+        }),
+      ).toEqual(
+        testUserStatusState.set(makeUserId(3), {
+          away: true,
+          status_text: null,
+          status_emoji: null,
+        }),
+      );
+    });
+
+    test('if the user already has user status stored update their key', () => {
+      expect(
+        userStatusesReducer(
+          Immutable.Map([[makeUserId(1), { away: false, status_text: 'foo', status_emoji: null }]]),
+          {
+            id: 0,
+            type: EVENT_USER_STATUS_UPDATE,
+            user_id: makeUserId(1),
+            status_text: 'bar',
+          },
+        ),
+      ).toEqual(Immutable.Map([[1, { away: false, status_text: 'bar', status_emoji: null }]]));
+    });
+
+    test('when the user_ status text is updated this is reflected in the state', () => {
+      expect(
+        userStatusesReducer(Immutable.Map(), {
+          id: 0,
+          type: EVENT_USER_STATUS_UPDATE,
+          user_id: makeUserId(1),
+          status_text: 'Hello, world!',
+        }),
```

Comment:
One thing you could do to help make these test cases less tedious to write (to help with then writing some new ones) is a `mkAction`-type helper -- in fact probably just taking a `UserStatusUpdate` as argument, and always using `eg.selfUser`.

Reply:
> and always using `eg.selfUser`

It often ends up being important to coordinate the IDs between the state objects (initial and expected) and the action. Maybe good to have a helper for constructing state objects? (And maybe that helper should use the reducer itself, like how we make `baseState` in `unreadModel-test.js`?)

Reply:
Yeah, possibly. Though I think these tests are fine as they are, in the new revision -- so further abstraction like that would mainly be useful if we were adding further complexity to this model, calling for more tests.

One thing that's going on in unreadModel-test.js, particularly in the EVENT_UPDATE_MESSAGE and EVENT_UPDATE_MESSAGE_FLAGS sections, is that:
* there's one kind of update that's very straightforward to implement -- adding a new message with EVENT_NEW_MESSAGE -- and can be used to build up an arbitrary state;
* then there's some other kinds of updates -- those other two event types -- where the implementation is much more complex, requiring careful testing.

So that first, very straightforward type of event makes a great way to build up a test state for use in testing the other types.

This is also a bit like how in the message-list diffing tests we use the fact that diffing from empty to a given list, and applying such a diff, are much simpler than making and applying a diff between two arbitrary lists. So to check if we've correctly handled an arbitrary diff, we make and apply those two much simpler diffs to get the before and after pictures of the DOM.

Here in the user-status model, there isn't so much of a gap between different levels of complexity. The thing we do when an update comes in is rather more complex than the unread model's new-message case… and then the only other nontrivial thing we do is just that exact same logic in a loop.

Reply:
> Here in the user-status model, there isn't so much of a gap between different levels of complexity.

Interesting; thanks for this helpful reasoning!